// pyano_local_modules/util/Alert.js
const Swal = require("sweetalert2");
let smallMixin = Swal.mixin({
	animation: false,
	customClass: 'animated fadeIn',
	position: "bottom-start",
	showConfirmButton: false,
	timer: 4000,
	toast: true,

});
const withConfirm = {
	cancelButtonText: "No",
	confirmButtonText: "Yes",
	showCancelButton: true,
	showConfirmButton: true,
	timer: null,

};

const blockingOptions = {
	allowEnterKey: false,
	allowEscapeKey: false,
	allowOutsideClick: false,
	animation: false,
	customClass: 'animated fadeIn',
	showCancelButton: false, // default false
	showCloseButton: false, // default false
	showConfirmButton: false,

};
const blockingSwalMixin = Swal.mixin(blockingOptions);

const Alert = (() => {
	const small = (() => {
		const info = (title, text = null, showConfirmBtns = false) => {
			let infoOptions = {
				title,
				text,
				type: "info",
			};
			if (showConfirmBtns)
				infoOptions = { ...infoOptions, ...withConfirm };
			return smallMixin.fire(infoOptions);
		};
		const _info = options => smallMixin.fire({ ...options, ...{ type: 'info' } });

		const success = (title, text = null, timer = 4000) =>
			smallMixin.fire({
				title,
				text,
				type: "success",
				timer
			});
		const warning = (title, text = null, showConfirmBtns = false) => {
			let warningOptions = {
				title,
				text,
				type: "warning"
			};
			if (showConfirmBtns)
				warningOptions = { ...warningOptions, ...withConfirm };
			return smallMixin.fire(warningOptions);
		};
		const _warning = options => smallMixin.fire({ ...options, ...{ type: 'warning' } });
		const warnThen = (title, text = null, onAfterClose = {}) =>
			smallMixin.fire({
				title,
				text,
				type: "warning",
				onAfterClose,
			});
		const error = (title, text) =>
			smallMixin.fire({
				title,
				text,
				type: "error",

			});
		return { info, _info, success, warning, _warning, warnThen, error };
	})();
	const big = (() => {
		// TODO: unused
		const basic = (title, text, onAfterClose = {}, confirmButtonText = "Done") =>
			Swal.fire({
				allowEnterKey: false,
				allowEscapeKey: false,
				allowOutsideClick: false,
				confirmButtonText,
				onAfterClose,
				preConfirm: onAfterClose,
				showConfirmButton: false,
				text,
				title,
			});
		const warning = options => blockingSwalMixin.fire({ showConfirmButton: true, ...options });
		/** @param options
		 @param {String[]?} strings
		 @param {function?} clickFn*/
		const blocking = (options, { strings, clickFn } = {}) => {
			if (strings && clickFn) {
				const $ = require("jquery");
				strings = strings
					.map(s => $(`<p class="clickable">${s}</p>`))
					.map($s => $s.click(() => clickFn($s)));
				options = {
					...options,
					onBeforeOpen: () => $('#swal2-content')
						.show()
						.append(strings)
				};
			} else {
				options = {
					...options,
					showConfirmButton: true,
					showCancelButton: true,
				};
			}
			return Swal.fire({ ...blockingOptions, ...options });
		};
		return { warning, blocking };
	})();
	return { small, big };
})();


module.exports = { Alert, blockingSwalMixin, smallMixin };
