// pyano_local_modules/util/Alert.js
const Swal = require("sweetalert2");
let smallMixin = Swal.mixin({
	position: "bottom-start",
	toast: true,
	animation: false,
	customClass: 'animated fadeIn',
	timer: 4000,
	showConfirmButton: false,

});
const withConfirm = {
	showCancelButton: true,
	timer: null,
	showConfirmButton: true,
	confirmButtonText: "Yes",
	cancelButtonText: "No"

};
const Alert = (() => {
	const small = (() => {
		const info = (title, text = null, showConfirmBtns = false) => {
			let infoOptions = {
				title,
				text,
				type: "info",
			};
			if (showConfirmBtns)
				infoOptions = { ...infoOptions, ...withConfirm };
			return smallMixin.fire(infoOptions);
		};


		const success = (title, text = null, timer = 4000) =>
			smallMixin.fire({
				title,
				text,
				type: "success",
				timer
			});
		const warning = (title, text = null, showConfirmBtns = false) => {
			let warningOptions = {
				title,
				text,
				type: "warning"
			};
			if (showConfirmBtns)
				warningOptions = { ...warningOptions, ...withConfirm };
			return smallMixin.fire(warningOptions);
		};
		const warnThen = (title, text = null, onAfterClose = {}) =>
			smallMixin.fire({
				title,
				text,
				type: "warning",
				onAfterClose,
			});
		const error = (title, text) =>
			smallMixin.fire({
				title,
				text,
				type: "error",

			});
		return { info, success, warning, warnThen, error };
	})();
	const big = (() => {
		const basic = (title, text, onAfterClose = {}, confirmButtonText = "Done") =>
			Swal.fire({
				title,
				text,
				onAfterClose,
				preConfirm: onAfterClose,
				allowOutsideClick: false,
				allowEscapeKey: false,
				allowEnterKey: false,
				showConfirmButton: false,
				confirmButtonText
			});
		const warning = (title, text) => Swal.fire({
			title,
			text,
			allowOutsideClick: false,
			allowEscapeKey: false,
			allowEnterKey: false,
			showConfirmButton: false,
		});
		return { basic, warning };
	})();
	return { small, big };
})();

/**@type {Swal}*/
const blockingSwalMixin = Swal.mixin({
	allowEnterKey: false,
	allowEscapeKey: false,
	allowOutsideClick: false,
	showCancelButton: false,
	showCloseButton: false,
	showConfirmButton: false,
	animation: false,

	// backdrop: false,
	customClass: 'animated fadeIn',

});

module.exports = { Alert, blockingSwalMixin, smallMixin };
