// pyano_local_modules/pages/VideoTools/videotools.js
const log = s => [`%c${s}`, 'color: #245ed1'];
let { EStore, Python } = require("pyano_local_modules/ext_libs");
let { zip, round, path_exists, remove_ext, $fadeOut, $fadeInMany, $fadeOutMany, $fadeIn } = require("pyano_local_modules/util");
let { safeSwitchCss, $MainContent, $Sidebar, $Title } = require("pyano_local_modules/document");
let { Alert } = require("pyano_local_modules/util/Alert");
let Path = require("path");
const $ = require('jquery');
const fs = require("fs");

/**@return {Number[]}*/
async function getOnsetTimes(filePath) {
	console.log(...log(`getOnsetTimes(${filePath})`));
	Alert.small.info('Getting onsets...');
	const options = { args: [filePath] };
	let onsets = await Python.runAsync("Record/get_onset_times.py", options);
	console.log(...log('onsets'), onsets);
	return onsets;


}

/**@param {String} filePath
 @param {Number} firstOnset
 @param {Number} silenceLen
 @param {Number?} toSecs
 @param {Number?} notesNum*/
async function trimAndForceSilence(filePath, firstOnset, silenceLen, toSecs, notesNum) {
	console.log(...log(`trimAndForceSilence(filePath, ${firstOnset}, ${silenceLen}, ${toSecs}, ${notesNum})`));
	let options = { args: [filePath, firstOnset, silenceLen] };
	if (toSecs && notesNum)
		options.args.push(toSecs, notesNum);
	const trimmedFilePath = await Python.runAsync("Record/trim_and_force_silence.py", options);
	console.log(...log('\ttrimmedFilePath'), trimmedFilePath);
	return trimmedFilePath[0];


}

async function handleMov(name, path) {
	let { value: shouldConvert } = await Alert.big.blocking({
		title: 'Convert to mp4?',
		html: name,
		showConfirmButton: true,
		showCancelButton: true,
	});
	if (!shouldConvert) {
		Alert.small.info('Not converting.');
		return;
	}

	let { execSync } = require('child_process');
	let mp4 = Path.join(Path.dirname(path), Path.basename(name, '.mov')) + '.mp4';
	try {
		execSync(`ffmpeg -i "${path}" -codec copy "${mp4}"`, { encoding: 'utf8', stdio: 'inherit' });
		Alert.small.success('Converted successfully.');
	} catch (e) {
		Alert.small.error('Convert failed', 'Maybe the converted file exists already?');
	}


}

const MP4 = (() => {
	async function createSubVersions(trimmedName, path) {
		console.log(...log(`createSubVersions(trimmedName = ${trimmedName}, path)`));
		let match = EStore.truthFilesList('txt')
		                  .find(f => remove_ext(f) == trimmedName.upTo('_trimmed'));
		if (!match) {

			Alert.small.error(`Didn't find a matching truth file`, `Tried to match: ${trimmedName.upTo('_trimmed')}`);
			return;
		}
		let shouldContinue = await Alert.big.blocking({
			title: `Found a matching truth file`, html: `Video will be cut based on ${match}`,
			showConfirmButton: true,
			showCancelButton: true
		}).value;
		if (!shouldContinue) {
			Alert.small.info('Aborting.');
			return;
		}
		const silenceLen = EStore.vidSilenceLen();
		let levels = EStore.currentTest().levels;
		let notes = levels.map(level => level.notes).rsort();
		shouldContinue = await Alert.big.blocking({
			title: `${notes.length} videos will be created`,
			html: `respective of these notes: ${notes}`,
			showConfirmButton: true,
			showCancelButton: true
		}).value;
		if (!shouldContinue) {
			Alert.small.info('Aborting.');
			return;
		}
		let truthFilePathNoExt = EStore.truthFilePath();
		let trimmedFilePath = `${truthFilePathNoExt}_trimmed.mp4`;
		console.log(...log(`\tcreateSubVersions`), { notes, truthFilePathNoExt, trimmedFilePath });
		let messages = await Python.runAsync('InsideTest/normalize_txt_file.py', { args: [`${truthFilePathNoExt}.txt`], mode: "json" });
		console.log(...log(`\tafter normalize_txt_file.py`), { messages });
		let tdeltas = messages.slice(0, notes[0]).map(msg => msg.time_delta - 0.01);
		if (notes[0] >= messages.length) {
			// if user specified as much notes as in txt file, add a fake "last note" lasting 1 second
			// this is to prevent cropping of last note
			tdeltas.push(1);
		}
		console.log({ tdeltas });
		let vidsDurations = notes.map(notesCount => tdeltas.slice(0, notesCount + 1).reduce((n, m) => n + m));
		console.log({ vidsDurations });
		let existed = 0;
		for (let [noteCount, vidDur] of zip(notes, vidsDurations)) {
			try {
				let trimmedPath = await trimAndForceSilence(trimmedFilePath, 0, silenceLen, vidDur, noteCount);
				console.log({ trimmedPath });
			} catch (e) {
				if (RegExp("^FileExistsError:.*.mp4$").test(e.message))
					existed += 1;
			}
		}
		let title = `Created ${notes.length - existed} videos`;
		if (existed != 0)
			title += ` (${existed} already existed)`;
		Alert.small.success(title, `of file: ${truthFilePathNoExt}`);
	}

	async function cleanSilence(name, path) {
		console.log(...log(`cleanSilence(name = ${name}, path)`));
		let { value: shouldClean } = await Alert.big.blocking({
			title: 'Clean silence from start?',
			html: name,
			showConfirmButton: true,
			showCancelButton: true,
		});
		if (!shouldClean) {
			Alert.small.info('Not cleaning.');
			return;
		}

		let trimmedFilePath = `${remove_ext(path)}_trimmed.mp4`;
		let trimmedExists = await path_exists(trimmedFilePath);
		console.log({ trimmedFilePath, trimmedExists });
		if (trimmedExists) {
			let { value: doAnyway } = await Alert.small.warning("A trimmed version of this file already exists. Continue anyway?", `Here: ${trimmedFilePath}`, true);
			console.log({ doAnyway });
			if (doAnyway) {
				await fs.unlinkSync(trimmedFilePath);
			} else {
				Alert.small.info('Not trimming.');
				return;
			}
		}

		const onsets = await getOnsetTimes(path);
		Alert.big.blocking({ title: 'How long was it until you played the first note?' },
			{
				strings: onsets.map(o => round(o, 1)),
				clickFn: async $s => {
					let secs = $s.text();
					Alert.small.info(`Trimming ${secs}s of silence...`);
					await trimAndForceSilence(path, secs, EStore.vidSilenceLen());
					Alert.small.success(`Trimmed ${secs}s successfully.`);
				}
			});


	}

	return { cleanSilence, createSubVersions };
})();

async function handleMp4(name, path) {
	if ('trimmed'.in(name)) {
		await MP4.createSubVersions(name, path);
	} else {
		await MP4.cleanSilence(name, path);
	}

}


async function onDrop(e) {
	let { name, path, size, type } = e.originalEvent.dataTransfer.files[0];
	if (!"video".in(type)) {
		Alert.small.warning('Not a video file');
		return;
	}
	name = name.lower();
	if (name.endsWith('mov'))
		return await handleMov(name, path);
	if (name.endsWith('mp4'))
		return await handleMp4(name, path);
}

const videoToolsPage = {
	switch: async (reload) => {
		console.log(...log(`videoToolsPage.switch(${reload})`));
		EStore.setLastPage('video_tools');
		if (reload) {
			let { remote } = require("electron");
			const currentWindow = remote.getCurrentWindow();
			currentWindow.reload();
		}
		await $fadeOut($MainContent, 100);
		$MainContent.empty();
		require("pyano_local_modules/sidebar").to_video_tools();
		safeSwitchCss("templates/css/video_tools.css");
		let $bigMessage = $('<div id="big_message">')
			.text('Drop a video file');
		let $dropArea = $('<div id="drop_area">')
			.on({
				'dragover dragenter': e => {
					e.preventDefault();
					e.stopPropagation();
				}, 'drop': async e => await onDrop(e)
			});

		$MainContent.append($bigMessage, $dropArea);
		await $fadeIn($MainContent, 300);
	}
};

module.exports = videoToolsPage;
