// pyano_local_modules/pages/InsideTest/Gui.js
// **pages/InsideTest/Gui.js
const $ = require('jquery');
let {
	concurrent, $fadeOut, $fadeIn, int, wait, num2eng, strong, range, $fadeOutMany, $fadeInMany, bool
} = require("pyano_local_modules/util");
let { EStore } = require("pyano_local_modules/ext_libs");
const Svg = require("./Svg");


class Video {
	/**@param {jQuery} $video
	 @param {String} vidsrcfile*/
	constructor($video, vidsrcfile) {
		this.$element = $video;
		this.src = `<source src="${vidsrcfile}" type="video/mp4">`;
		this.$element.append(this.src);
	}

	_promiseClick() {
		console.log('waiting for button click');
		return new Promise(resolve => {
			document.body.addEventListener("click", resolve, { once: true });
		});
	}

	/**@return {HTMLMediaElement}*/
	get domElm() {return this.$element[0];}


	/**@return {Promise<void>}*/
	async play() {


		try {
			await concurrent(this.domElm.play(), wait(this.domElm.duration * 1000));
		} catch (e) {
			console.warn(e);
			let msg = await this._promiseClick();
			console.log(msg);
			await concurrent(this.domElm.play(), wait(this.domElm.duration * 1000));
		}

	}

	async pause() {
		return await this.domElm.pause();
	}
}

class Animation {
	/**@param {jQuery} $animation*/
	constructor($animation) {
		/**@type {jQuery}*/
		this.$element = $animation;
		let svgs = [];
		for (let i of range(0, 51)) {
			if (i == 0) {
				svgs.push(Svg.blackRight);
				continue;
			} else if (i == 51) {
				svgs.push(Svg.base);
				continue;
			}
			let searchElement = i % 7;
			if ([2, 5].includes(searchElement))
				svgs.push(Svg.blackRight);
			else if ([1, 4, 8].includes(searchElement))
				svgs.push(Svg.blackLeft);
			else // [0, 3, 6, 7]
				svgs.push(Svg.blackBoth);
		}
		this.$element.append(svgs);

	}

	/**@param {Number} note*/
	_note2svgIndex(note) {
		let scaleStart = int(note / 12) * 7 - 11;
		let indexInScale = note % 12;
		let index = scaleStart + [0, 0, 1, 1, 2, 3, 3, 4, 4, 5, 5, 6][indexInScale];
		if (![0, 2, 4, 5, 7, 9, 11].includes(indexInScale))
			return [index, index + 1];
		return index;
		/*let trans = {
			0: 0,
			1: 0.5,
			2: 1,
			3: 2,
			4: 2.5,
			5: 3,
			6: 3.5,
			7: 4,
			8: 5,
			9: 5.5,
			10: 6,
			11: 6.5,
		};

		note -= 9;
		let index = (int(note / 12) * 7) + trans[note % 12];
		return parseFloat(index);
		*/
	}

	/**
	 @param {Number} note
	 @param {Number} noteDuration
	 @param {String} fill
	 * */
	async markKey(note, noteDuration, fill) {
		return new Promise(async resolve => {
			let index = this._note2svgIndex(note);
			let $toMark, baseFill;
			let $children = this.$element.children();
			if (index.length != undefined) { // [3,4] for example
				let [indexLeft, indexRight] = index;
				let $keyLeft = $children.filter(`svg:nth-child(${indexRight})`).children(`.piano-black-left-rect`);
				let $keyRight = $children.filter(`svg:nth-child(${indexLeft})`).children(`.piano-black-right-rect`);
				$toMark = $keyLeft.add($keyRight);
				baseFill = 'black';
			} else { // 3
				$toMark = $children.filter(`svg:nth-child(${index})`).children(`.piano-white-rect`);
				baseFill = 'white';
			}
			$toMark.css({ fill });
			await wait(noteDuration);
			$toMark.css({ fill: baseFill });
			resolve();
			// setTimeout(() => {
			//
			// }, noteDuration);

		});

	}


}

/**@param {Midi} midi
 @param {Piano|Tone.AudioNode} playbackPiano*/
async function playWholeTruthDemo(midi, playbackPiano) {
	console.log('%cGui.playWholeTruthDemo()', 'color:#9876aa');
	$bigMessage.text(`A tutorial`);
	await $fadeIn($bigMessage, 700);
	let smallMessage = `Here's a demo that shows you what you will be playing.`;
	$smallMessage.html(smallMessage);
	await $fadeIn($smallMessage, 700);
	await wait(1500);
	await $fadeOutMany(1500, $smallMessage, $bigMessage);

	if (EStore.currentTest().demo_type == "animation") {
		await $fadeIn(animation.$element, 700);
		await midi.playMidiFile({ midiFilePath: EStore.truthFilePath('mid'), animation, playbackPiano });
		await wait(1500);
		await $fadeOut(animation.$element, 700);
	} else {
		await $fadeIn(video.$element, 700);
		await video.play();
		await wait(1500);
		await $fadeOut(video.$element, 700);
	}

}


/**@param {Midi} midi
 @param {Piano} playbackPiano
 @param {Number} levelIndex
 @param {Number} trialIndex
 @param {Number} numOfNotes
 @return {Promise<void>}*/
async function playPreTrialDemo(midi, playbackPiano, levelIndex, trialIndex, numOfNotes) {
	console.log('%cGui.playPreTrialDemo()', 'color:#9876aa');
	const trialNumEnglish = num2eng(trialIndex + 1, true);
	if (trialIndex == 0)
		$bigMessage.text(`${num2eng(levelIndex + 1, true).title()} level, ${trialNumEnglish} trial.`);
	else
		$bigMessage.text(`Let's move on to the ${trialNumEnglish} trial.`);
	await $fadeIn($bigMessage, 700);

	$smallMessage.html(`You are going to play <b>${numOfNotes}</b> notes.`);
	await $fadeIn($smallMessage, 700);
	await wait(700);
	$smallMessageSecondary.html(`Here's an animation showing only these <b>${numOfNotes}</b> notes.`);
	await $fadeIn($smallMessageSecondary, 700);
	await wait(1500);
	await $fadeOutMany(1500, $bigMessage, $smallMessage, $smallMessageSecondary);
	await $fadeIn(animation.$element, 700);
	await midi.playMidiFile({ midiFilePath: EStore.truthFilePath('mid'), animation, numOfNotes, playbackPiano });
	await wait(1500);
	await $fadeOut(animation.$element, 700);
	$smallMessage.html(`Please play <b>${numOfNotes}</b> notes.`);
	await $fadeInMany(300, $bigButton, $smallMessage);

}


/**@param {Midi} midi
 @param {Piano} playbackPiano
 @param {Number} trialIndex
 @param {String[]} mistakes
 @return {Promise<void>}*/
async function showFailedTrialFeedback(midi, playbackPiano, trialIndex, mistakes) {
	console.log('%cGui.showFailedTrialFeedback()', 'color:#9876aa');
	$bigMessage.text("Oi vei!");
	await concurrent($fadeIn($bigMessage, 700), $fadeOutMany(200, $smallMessage, $bigButton));
	await wait(700);
	// await $fadeOutMany(200, $smallMessage, $bigButton);
	// await $fadeIn($bigMessage, 700);
	const badAccuracyInTempo = bool(mistakes)
	                           && 'note'.in(mistakes)
	                           && EStore.isInTempo();
	if (badAccuracyInTempo) {
		$smallMessage.text(`You got some notes wrong.`);
		$smallMessageSecondary.text("Please try this trial again.");
	} else {
		$smallMessage.text(`You almost got the ${num2eng(trialIndex + 1)} trial right.`);
		$smallMessageSecondary.text("Here's an animation showing the errors.");
	}
	await $fadeIn($smallMessage, 700);
	await wait(700);
	await $fadeIn($smallMessageSecondary, 700);
	await wait(1500);
	await $fadeOutMany(1500, $bigMessage, $smallMessage, $smallMessageSecondary);
	await $fadeIn(animation.$element, 700);
	const numOfNotes = mistakes.length;

	let playOptions = {
		midiFilePath: EStore.truthFilePath('mid'),
		animation,
		numOfNotes,
		mistakes,
		playbackPiano
	};
	if (EStore.isInTempo())
		playOptions.speed = EStore.currentTest().errors_playingspeed;
	await midi.playMidiFile(playOptions);
	await wait(1500);
	await $fadeOut(animation.$element, 700);
}

/**@param {Midi} midi
 @param {Piano} playbackPiano
 @param {Number} trialIndex*/
async function showPassedTrialFeedback(midi, playbackPiano, trialIndex) {
	console.log('%cGui.showPassedTrialFeedback()', 'color:#9876aa');
	$bigMessage.text(`Good job! You passed the ${num2eng(trialIndex + 1)} trial.`);
	await concurrent($fadeIn($bigMessage, 700), $fadeOutMany(200, $smallMessage, $bigButton));
	await wait(700);
	const numOfNotes = EStore.numOfNotesCurrLevel();
	$smallMessage.html(`Let's show you those same <b>${numOfNotes}</b> notes you just got correctly.`);
	$smallMessageSecondary.html(`This will make you even better.`);
	await $fadeIn($smallMessage, 700);
	await wait(700);
	await $fadeIn($smallMessageSecondary, 700);
	await wait(1500);
	await $fadeOutMany(1000, $bigMessage, $smallMessage, $smallMessageSecondary);
	await $fadeIn(animation.$element, 700);
	await midi.playMidiFile({
		midiFilePath: EStore.truthFilePath('mid'),
		animation,
		numOfNotes,
		playbackPiano
	});
	await wait(1000);
	await $fadeOut(animation.$element, 700);
}

async function showTestCompleteMessages() {
	console.log('%cGui.showTestCompleteMessages()', 'color:#9876aa');
	$bigMessage.text(`Test is over!`);
	$smallMessage.html('You deserve an ice cream. ðŸ¦');

	await $fadeInMany(700, $bigMessage, $smallMessage);
}


/**@param {TLevel[]} levels
 @return {Number[]}*/
function updateLevelTrialSubtitles(levels) {
	let [levelIndex, trialIndex] = EStore.currentTrialCoords();
	$('#page_subtitle > div:nth-child(1) > strong')
		.text(EStore.truthFileName({ withExtension: false }));
	$('#page_subtitle > div:nth-child(2) > strong')
		.text(`${levelIndex + 1}/${levels.length}`);
	$('#page_subtitle > div:nth-child(3) > strong')
		.text(`${trialIndex + 1}/${levels[levelIndex].trials}`);

	return [levelIndex, trialIndex];
}

let $pageSubtitle = $('<div class="subtitle" id="page_subtitle">')
	.append(
		`<div>Truth file: ${strong(EStore.truthFileName({ withExtension: false }))}</div>`,
		`<div>Level: <strong>1/1</strong></div>`,
		`<div>Trial: <strong>1/1</strong></div>`,
	);
let $bigMessage = $('<div id="big_message">').hide();
let $smallMessage = $('<div class="subtitle" id="small_message">').hide();
let $smallMessageSecondary = $('<div class="subtitle" id="small_message_secondary">')
	.hide();
let $bigButton = $('<button id="big_button">')
	.addClass("my-btn active-btn")
	.append("<span>I'm done playing</span>")
	.hide();
let animation = new Animation($('<div id="animation">').hide());
let vidsrcfile = `file:///${EStore.truthFilePath('mp4')}`;
let video = new Video($(`<video id="video">`).hide(), vidsrcfile);


module.exports = {
	$pageSubtitle,
	$bigMessage,
	$smallMessageSecondary,
	$smallMessage,
	$bigButton,
	playPreTrialDemo,
	showPassedTrialFeedback,
	showFailedTrialFeedback,
	showTestCompleteMessages,
	playWholeTruthDemo,
	updateLevelTrialSubtitles,
	animation,
	video
};
