// pyano_local_modules/ext_libs/store.js
let Store = require('electron-store');
let Path = require("path");
const fs = require('fs');

class MyStore extends Store {
	constructor() {
		super();
		this._doTruthFileCheck();

	}

	/**@private*/
	_doTruthFileCheck() {
		console.log('MyStore._doTruthFileCheck()');
		const truth = this.truth();
		truth.txt.allExist()
		     .then(async exist => {
			     if (exist) {
				     return Alert.small.success(`All "${truth.name}" txt files exist.`);
			     } else {
				     let txtFilesList = this.truthFilesList('txt').map(fsx.remove_ext);
				     let filteredTxts = txtFilesList.filter(a => txtFilesList.filter(txt => txt.startsWith(a)).length >= 3);
				     if (!bool(filteredTxts))
					     return await Alert.big.warning({
						     title: 'No valid truth files found',
						     html: 'There needs to be at least one txt file with 2 "on" and "off" counterparts.'
					     });


				     await Alert.big.blocking({
					     title: `Truth file invalid: ${truth.name}`,
					     html: '<b>Please choose one of the following valid truths:</b>',
				     }, {
					     strings: filteredTxts,
					     clickFn: async $s => {
						     try {
							     await this.setTruthFilePath(new Truth(Path.join(this.truthsDirPath(), $s.text())));
							     this.set({
								     'current_test.finished_trials_count': 0,
								     'current_test.levels': []
							     });
							     reloadPage();
						     } catch (err) {
							     document.getElementById('swal2-title').innerText = err.message;
							     document.getElementById('swal2-content').style.display = 'none';
							     document.getElementsByClassName('swal2-icon swal2-warning')[0].style.display = 'inherit';
							     throw err;
						     }

					     }
				     });
			     }
		     });

	}

	/**@return {TCurrentTest}*/
	currentTest() {
		return this.get('current_test');
	}

	/**@param {Truth} truth
	 @returns {Promise<Truth>}*/
	setTruthFilePath(truth) {
		return truth.txt.allExist()
		            .then(exist => {
			            if (exist) {
				            this.set('current_test.truth_file_path', `experiments\\truths\\${truth.txt.base.name}`);
				            return truth;
			            } else {
				            throw new Error(`Not all txt files of truth exist: ${truth.txt.base.name}`);
			            }
		            });
	}

	/**@param {TDemoType} type*/
	setDemoType(type) {
		if (!type.in(['video', 'animation']))
			throw new Error(`setDemoMode(type = ${type}), can be either video or animation`);
		this.set('current_test.demo_type', type);
	}

	/**@param {TLastPage} page*/
	setLastPage(page) {
		if (!page.in(['new_test', 'inside_test', 'record', 'file_tools', 'settings']))
			throw new Error(`setLastPage(page = ${page}), can be either ['new_test','inside_test','record','file_tools','settings']`);

		this.set('last_page', page);
	}

	/**@example
	 update('subjects', [names])
	 @param {string} K
	 @param kv
	 @return {*} */
	update(K, kv) {
		let V = this.get(K);
		if (Array.isArray(V)) {
			this.set(K, [...V, kv]);
		} else {
			Object.assign(V, kv);
			this.set(K, V);
		}
		return this.get(K);
	}

	/** @param {string} K
	 @return {TCurrentTest} */
	increase(K) {
		let V = this.get(K);
		this.set(K, V + 1);
		return this.get('current_test');
	}

	rootAbsPath() {
		return this.get('root_abs_path');
	}

	vidSilenceLen() {
		return this.get('vid_silence_len');
	}

	/**"C:\Sync\Code\Python\Pyano\pyano_01\src\experiments\truths"
	 @return {string}*/
	truthsDirPath() {
		return Path.join(this.rootAbsPath(), 'experiments', 'truths');
	}

	/**@param {String?} extFilter
	 @return {string[]} truthFiles*/
	truthFilesList(extFilter = null) {
		if (extFilter != null)
			if (!extFilter.in(['txt', 'mid', 'mp4']))
				throw new Error(`truthFilesList(extFilter = ${extFilter}), must be either ['txt','mid','mp4'] or not at all`);

		const truthsDirPath = this.truthsDirPath();

		let truthFiles = [...new Set(fs.readdirSync(truthsDirPath))];
		if (extFilter != null)
			return truthFiles.filter(f => Path.extname(f) == `.${extFilter}`);
		return truthFiles;
	}

	/**"my_truth[.txt]"
	 @example
	 truthFileName()
	 >>> my_truth
	 truthFileName({withExtension:true})
	 >>> my_truth.txt
	 @param {boolean} withExtension
	 @return {string}*/
	truthFileName({ withExtension }) {

		const truthFilePath = this.get("current_test.truth_file_path");
		if (withExtension)
			return fsx.basename(truthFilePath);
		else
			return fsx.basename(truthFilePath, '.txt');
	}


	/**"C:\Sync\Code\Python\Pyano\pyano_01\src\experiments\truths\my_truth[.?]"
	 @example
	 truthFilePath()
	 >>> "C:\Sync\Code\Python\Pyano\pyano_01\src\experiments\truths\my_truth"
	 truthFilePath('mid')
	 >>> "C:\Sync\Code\Python\Pyano\pyano_01\src\experiments\truths\my_truth.mid"
	 @param {String?} ext
	 @return {string}*/
	truthFilePath(ext = null) {
		let truthFileName = this.truthFileName({ withExtension: false });
		if (ext != null) {
			if (ext.includes('.'))
				throw new Error(`ext included dot. Pass only file type. received: ${ext}`);
			if (!ext.in(['txt', 'mid', 'mp4']))
				throw new Error(`ext is not txt, mid or mp4. Pass only one of these. received: ${ext}`);

			return Path.join(this.truthsDirPath(), `${truthFileName}.${ext}`);
		} else
			return Path.join(this.truthsDirPath(), `${truthFileName}`);
	}

	/**@return {Truth}*/
	truth() {
		return new Truth(Path.join(this.truthsDirPath(), Path.basename(this.get("current_test.truth_file_path"), '.txt')));
	}

	/** "C:\Sync\Code\Python\Pyano\pyano_01\src\experiments\subjects"
	 @return {string} */
	subjectsDirPath() {
		return Path.join(this.rootAbsPath(), 'experiments', 'subjects');
	}

	/** "gilad"
	 @return {string} */
	subjectName() {
		return this.get('current_test.current_subject');
	}

	/** "C:\Sync\Code\Python\Pyano\pyano_01\src\experiments\subjects\gilad"
	 @return {string} */
	subjectDirPath() {
		return Path.join(this.subjectsDirPath(), this.subjectName());
	}

	salamanderDirPath() {
		return Path.join(this.rootAbsPath(), 'templates', 'Salamander/');
	}

	/**"c:\Sync\Code\Python\Pyano\pyano_01\src\experiments\subjects\gilad\fur_elise_accuracy"
	 @return {string}
	 */
	testOutPath() {
		const currSubjectDir = this.subjectDirPath(); // ".../subjects/gilad"
		const truthFileName = this.truthFileName({ withExtension: false }); // "fur_elise"
		const { learning_type } = this.get('current_test'); // "accuracy"
		return Path.join(currSubjectDir, `${truthFileName}_${learning_type}`);
	}

	/**"c:\Sync\Code\Python\Pyano\pyano_01\src\experiments\subjects\gilad\fur_elise_accuracy\level_0_trial_0[.?]"
	 @example
	 trialFilePath()
	 >>> "c:\Sync\Code\Python\Pyano\pyano_01\src\experiments\subjects\gilad\fur_elise\level_0_trial_0"
	 trialFilePath('mid')
	 >>> "c:\Sync\Code\Python\Pyano\pyano_01\src\experiments\subjects\gilad\fur_elise\level_0_trial_0.mid"
	 @param {?string} [ext=null]
	 @return {string}*/
	trialFilePath(ext = null) {
		let [level_index, trial_index] = this.currentTrialCoords();
		const noExtension = Path.join(this.testOutPath(), `level_${level_index}_trial_${trial_index}`);
		if (ext == null)
			return noExtension;

		if (ext.includes('.'))
			throw new Error(`ext included dot. Pass only file type. received: ${ext}`);
		if (!ext.in(['txt', 'mid']))
			throw new Error(`ext is not txt or mid. Pass only one of these. received: ${ext}`);
		return `${noExtension}.${ext}`;

	}

	/**@return {Truth}*/
	trialTruth() {
		let [level_index, trial_index] = this.currentTrialCoords();
		return new Truth(Path.join(this.testOutPath(), `level_${level_index}_trial_${trial_index}`));
	}


	// ***DYNAMIC CONTENT

	/**@return {boolean}*/
	isLearningTempo() {
		throw new Error("LEARNING TYPE HAS BEEN DEPRECATED");
		return this.currentTest().learning_type == "tempo";
	}

	/**@return {boolean}*/
	isDemoVideo() {
		return this.currentTest().demo_type == "video";
	}

	/**@return {boolean}*/
	isWholeTestOver() {
		let { levels, finished_trials_count } = this.currentTest();
		return sum(levels.map(level => level.trials)) == finished_trials_count;
	}

	/**@return {number[]}*/
	currentTrialCoords() {
		let { levels, finished_trials_count } = this.currentTest();
		let flatTrialsList = levels.map(level => level.trials);
		for (let [levelIndex, trialsNum] of enumerate(flatTrialsList)) {

			let sumSoFar = sum(flatTrialsList.slice(0, levelIndex + 1));
			if (sumSoFar > finished_trials_count)
				return [levelIndex, trialsNum - (sumSoFar - finished_trials_count)];
		}
		throw "currentTrialCoords: out of index error";
	}

	/**@return {number}*/
	numOfNotesCurrLevel() {
		let level_index = this.currentTrialCoords()[0];
		return this.currentTest().levels[level_index].notes;
	}

	/**@return {number}*/
	maxNotesOfLevels() {
		return max(...this.currentTest().levels.map(lvl => lvl.notes));
	}

	/*	truthFileTotalNotes() {

			const truth = this.truth();
			try {
				return truth.numOfNotes();
			} catch (e) {}
		}
	*/
}

/**
 * @type {MyStore}
 */
module.exports = MyStore;
